sample-programs/erc20/erc20.scm

```diff
@@ -0,0 +1,129 @@
+
+;; Gets the amount of tokens owned by the specified address.
+(define (balance-of (account principal))
+  (let ((balance
+      (get balance 
+        (fetch-entry balances (tuple (owner account))))))
+    (if (eq? balance 'null) 0 balance)))
+
+;; Gets the amount of tokens that an owner allowed to a spender.
+(define (allowance-of (spender principal) (owner principal))
+  (let ((allowance (get allowance 
+      (fetch-entry allowances (tuple 
+        (owner owner) 
+        (spender spender))))))
+    (if (eq? allowance 'null) 0 allowance)))
+
+;; Credits balance of a specified principal.
+(define (credit-balance! (account principal) (amount int))
+  (if (<= amount 0)
+    'false
+    (let ((current-balance (balance-of account)))
+      (begin
+        (set-entry! balances 
+          (tuple (owner account))
+          (tuple (balance (+ amount current-balance)))) 
+        'true)))) ;; Overflow management?
+
+;; Debits balance of a specified principal.
+(define (debit-balance! (account principal) (amount int))
+  (let ((balance (balance-of account)))
+    (if (or (> amount balance) (<= amount 0))
+      'false
+      (begin
+        (set-entry! balances
+          (tuple (owner account))
+          (tuple (balance (- balance amount))))
+        'true))))
+
+;; Transfers tokens to a specified principal.
+(define (transfer! (sender principal) (recipient principal) (amount int))
+  (if (and (credit-balance! recipient amount) (debit-balance! sender amount)) 
+    'true
+    'false))
+
+;; Decrease allowance of a specified spender.
+(define (decrease-allowance! (spender principal) (owner principal) (amount int))
+  (let ((allowance (allowance-of spender owner)))
+    (if (or (> amount allowance) (<= amount 0))
+      'false
+      (begin
+        (set-entry! allowances
+          (tuple (spender spender) (owner owner))
+          (tuple (allowance (- allowance amount))))
+        'true))))
+
+;; Internal - Increase allowance of a specified spender.
+(define (increase-allowance! (spender principal) (owner principal) (amount int))
+  (let ((allowance (allowance-of spender owner)))
+    (if (<= amount 0)
+      'false
+      (begin
+        (set-entry! allowances 
+          (tuple (spender spender) (owner owner))
+          (tuple (allowance (+ allowance amount))))
+        'true)))) 
+
+;; Public functions
+
+;; Transfers tokens to a specified principal.
+(define-public (transfer (recipient principal) (amount int))
+  (transfer! tx-sender recipient amount))
+
+;; Transfers tokens to a specified principal, performed by a spender
+(define-public (transfer-from (owner principal) (recipient principal) (amount int))
+  (let ((allowance (allowance-of tx-sender owner)))
+    (if (or (> amount allowance) (<= amount 0))
+      'false
+      (begin
+        (transfer! owner recipient amount)
+        (decrease-allowance! tx-sender owner amount)))))
+
+;; Update the allowance for a given spender
+(define-public (approve (spender principal) (amount int))

```

Comment:
There aren't any checks on `spender`.  Is anyone allowed to approve any amount?

Reply:
Yes, as a token holder, you can designate an entity for spending an allowance that you define.
On a decentralized exchange, this gives you the ability to create a `make` order (vs `transfer`, giving you the ability to create a `take` order).